proposals/analytics.md

```diff
@@ -0,0 +1,58 @@
+# Analytics
+
+
+## Problem
+Breaking our monolithic application into several different applications has many benefits. It also introduces many levels of complexity.
+It becomes very difficult to manage complex dependencies like analytics configuration.
+This is exacerbated in the library context where, configuration, for example, must come from the application context.
+
+## Solution
+We need a cross library/application solution that we can inject configuration into at the application level.
+This solution must not leak into the component structure such that library dependencies become unusable without application configuration.
+
+### Initialisation
+In order to configure analytics we must build a library API that can be injected into:
+
+```js
+// module: utils/analytics
+
+import { init } from '@festicket/analytics';
+
+export const { track, identify }  = init({
+  ACCOUNT_KEY: process.env.ACCOUNT_KEY,
+  ACCOUNT_SECRET: process.env.ACCOUNT_SECRET,
+});
+```
+
+This means that, once configured, an application can identify or track with ease:
+
+```js
+// any file within the application
+import { track } from '@utils/analytics';
+
+// Do some stuff
+track({ type: 'MY_CUSTOM_EVENT', ...data });
```

Comment:
Will we just pass this object as-is to Segment, or do we want to abstract out an expected format here?

Reply:
I think, as a business, we are pretty sold on segment so we can probably just pass through in the format that it expects? I doubt an abstraction would bring much benefit here but would probably add an amount of complexity.

Honestly, I think if an application has programatic needs it will probably end up using something like [react-redux-segment](https://github.com/rangle/redux-segment). This is just an escape hatch so that you don't get tied into redux immediately, if you don't need it.

/cc @sam-at-work

Reply:
👍

Reply:
imo with redux it's super easy to handle analytics thru a middleware, either `react-redux-segment` or a custom `SFA` based like what we have in `festicket-app`. The tricky part here is to come up with a framework agnostic solution.

Reply:
The intention with the global click handle is that you have to have `data-analytics=true` for it to work, so it's **opt in**. If at an application level, you want to handle analytics in another way then you don't add the data attributes. To me, this is a flexible solution. Having it operate at a DOM level keeps it framework agnostic?

Reply:
Hmm, if you're having to add a new attribute to the link for it to be tracked, why not just add an onClick handler to call `track` directly? Then it's unified.

Reply:
no, what I meant is whatever is coupled to redux is not framework agnositc

Reply:
@dbbk because the `track` method is exposed on the **application level**. The header and footer, that need to fire track events on click, don't get to import the `track` function because they exist in the library context.

Reply:
@dbbk we tried that, eventually have to manually implement the tracking in every `onClick` became a mess quickly.

Reply:
@cutandpastey where will the logic live for a user signing in & signing out?
Won't that need to live in the `header` container, and therefore won't it need to be able to call `track` so we can log to segment both a successful and non-successful log in attempt?

Reply:
@sam-at-work Well, login should be done in a form and you can listen for the form submit event? In the case of the login form its quite specific so you can use that to your advantage? You could even just add the data attributes to the submit button?

Reply:
but you don't know if the log in was successful or not until we get the response from the server. 

And unless we do a full page reload when a user logs in, this solution would give us no way to correctly fire an `identify` call when the user logs in.

https://segment.com/docs/spec/identify/

Reply:
@sam-at-work depends on what you're tracking: if you want to track the form submitting, the login or both.

Reply:
@marcoala most importantly we need to be able to call `identify` after a successful login. https://segment.com/docs/spec/identify/

Reply:
I see, but still you can track submit and then `identify` after `submit`.

Reply:
@sam-at-work @marcoala That's exactly whats outlined here?:

https://github.com/festicket/front-end-meta/pull/2#discussion_r167881446